src/accelerate/data_loader.py

```diff
@@ -361,7 +361,15 @@ def __init__(self, dataset, device=None, rng_types=None, generator=None, **kwarg
 
     def __iter__(self):
         if self.rng_types is not None:
-            synchronize_rng_states(self.rng_types, self.generator)
+            batch_sampler = self.sampler if isinstance(self.sampler, BatchSampler) else self.batch_sampler
+            sampler = (
+                batch_sampler.batch_sampler.sampler
+                if hasattr(batch_sampler, "batch_sampler")
+                else batch_sampler.sampler
+            )
+            if hasattr(sampler, "generator"):
+                generator = sampler.generator
+                synchronize_rng_states(self.rng_types, generator)
         self.gradient_state._set_end_of_dataloader(False)
```

Comment:
We should have an else branch here in case `self.rng_types` contains something different than `"generator"`, by sending `None` for the generator argument (since we have removed `"generator"` from the rng_types at this stage, if there is none available.